Epic/37297 automatic meeting agendas and minutes linked to work packages

jjabari-op opened the pull request:
https://community.openproject.org/projects/openproject/work_packages/37297

jonasjabari committed:
include and setup hotwire js

jonasjabari committed:
introduce turbo_stream view_component abstraction

jonasjabari committed:
added simple demo for abstraction approach

klaustopher committed:
Fix Rubocop and ESLint errors

jjabari-op committed:
Merge pull request #13225 from opf/fix-hotwire-rubocop

Fix Rubocop and ESLint errors for Turbo abstraction

klaustopher committed:
Rename respond_with to respond_to_with and allow additional formats

jonasjabari committed:
fixed update action, which should only provide the inner html

jonasjabari committed:
switched to concern instead of base class

jonasjabari committed:
fixed module

jjabari-op committed:
[#37297] Automatic meeting agendas and minutes linked to work packages

https://community.openproject.org/work_packages/37297

jonasjabari committed:
check if component_wrapper is used and raise error if not

jonasjabari committed:
add alias method name

jjabari-op committed:
Merge branch 'introduce-hotwire-and-view-component' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

jjabari-op committed:
introducing flex and box collection component in order to enable pure ruby view components without erb

jjabari-op committed:
copied and adjusted code from prototypical spike branch as a starting point

jjabari-op committed:
added primer adjustment styles which fixes minor styling issues caused by OP styles overriding primer styles -> tbd

jjabari-op committed:
enabled more primer adjustments in order to match primer input field styles -> tbd

jjabari-op committed:
moved new form to bottom of list

jjabari-op committed:
adjusted prototypical implementation towards new mockups

jjabari-op committed:
added move actions via menu

jjabari-op committed:
added plus and grab icon

jjabari-op committed:
Merge branch 'dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

jjabari-op committed:
added dynamic notes section

jjabari-op committed:
Merge branch 'epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages' of github.com:opf/openproject into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

jjabari-op committed:
increase font size of notes as requested by parimal

jjabari-op committed:
added simple work_package reference to meeting_agenda_items

jjabari-op committed:
fixed edit state

jjabari-op committed:
hide move action if not necessary

jjabari-op committed:
switched to a default select until we have proper filterable single select ready

jjabari-op committed:
adding box slots when using flex direction defined by responsive breakpoints

jjabari-op committed:
added layout and editable header

jjabari-op committed:
added component_collection component in order to orchestrate multiple primer components withhout any wrapping boxes

jjabari-op committed:
added editable meeting details in sidebar

jjabari-op committed:
WIP: adding state section to sidebar

jjabari-op committed:
Merge branch 'dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

jjabari-op committed:
introducing open and closed meeting state

jjabari-op committed:
WIP: Adding basic participant management via dialog, buggy currently

jjabari-op committed:
temporarly fixed basic participants management by using non-primer checkboxes

jjabari-op committed:
fixed and finalized basic participant management

jjabari-op committed:
introducing basic backend permission checks and rendering actions only if allowed

jjabari-op committed:
removed confirmations in order to improve UX, cancel confirmation on agenda items with touched notes pending

jjabari-op committed:
added timezone display

jjabari-op committed:
truncated location display

jjabari-op committed:
fixing date formatting

jjabari-op committed:
localized strings

jjabari-op committed:
Merge branch 'dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

jjabari-op committed:
scope authorization to project roles

jjabari-op committed:
Merge branch 'epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages' of github.com:opf/openproject into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
Render op-autocompleter from rails

oliverguenther committed:
Replace flex page layout with Primer Layout

oliverguenther committed:
Make opAutocompleter a dynamic component

cbliard committed:
Make agenda item description a rich editor

jjabari-op committed:
introducing contracts and services for meeting_agenda_item actions

jjabari-op committed:
introduce prototypical flash messages and improve UX for users for stale browser windows

cbliard committed:
Manage error and value state

oliverguenther committed:
Bind to the whole object and load it when initially presented

oliverguenther committed:
Use markdown formatting for description

oliverguenther committed:
Parameterize options for autocomplete

oliverguenther committed:
Add author autocomplete field

oliverguenther committed:
Introduce primerized autocomplete styles

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
fixup! Make opAutocompleter a dynamic component

oliverguenther committed:
Add meeting type and restore standard meeting rendering

oliverguenther committed:
Set correct foreign_key to identify meeting

oliverguenther committed:
Don't render user avatar component if user not present

oliverguenther committed:
Add very basic crud spec for structured meetings

oliverguenther committed:
Linting

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
Linting

oliverguenther committed:
Fix resource in add-assignee component

oliverguenther committed:
Extend spec for structured meeting

oliverguenther committed:
Support for undisclosed linked work packages

jjabari-op committed:
adjusted time zone display as requested by Parimal

jjabari-op committed:
Merge branch 'epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages' of github.com:opf/openproject into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

jjabari-op committed:
adjusted duration display and input as requested by Parimal

jjabari-op committed:
further adjusted time zone display and input as request by Parimal

jjabari-op committed:
fixed permission mappings lost due to former merge conflict resolution

jjabari-op committed:
quick fix strong params access issue caused by the introduced meeting type, additional rubocop fixes

jjabari-op committed:
quickly added basic empty state

oliverguenther committed:
Introduce qa_selector finder and matcher to ease writing tests

oliverguenther committed:
Use qa_selector matcher

jjabari-op committed:
quick fixing focus issue cause by author autocomplete component, buggy -> not working for wp select, needs refactoring as this is just a quick fix

jjabari-op committed:
Merge branch 'epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages' of github.com:opf/openproject into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
fixup! Use qa_selector matcher

oliverguenther committed:
Rewrite spec usage with test_selector

HDinger committed:
Add project filter for author of MeetingAgendaItem

HDinger committed:
Allow Principals to be set as "author" of a MeetingAgendaItem (although the term is semantically wrong and repsonsible would fit better)

HDinger committed:
Show the name of the currently selected author in the select input field

HDinger committed:
Add test cases for agenda item author

oliverguenther committed:
Fix author specs

oliverguenther committed:
Linting

ulferts reviewed (changes requested):
I haven't yet taken a look into the code but only tried out the changes from a user perspective. Sorry for the barrage of topics. I generally like the change.

- [x] **Unclear purpose of the user selection**

Decision: We're removing the field in favor of an "added by" field

I believe this to be the consequence of not having [labels on those form fields](https://primer.style/design/ui-patterns/forms/overview#label-required)

- [x] **Possible to create an agenda item without user** (by design)

It is possible to have agenda items without a responsible users:

<img width="1014" alt="image" src="https://github.com/opf/openproject/assets/617519/eaa60d98-ff1f-421d-8c4a-a3e8105c0163">

This is because the responsible value can be deleted. Attempting to edit such an item will result in an error.

- [x] **Removing the last invitee from the meeting leads to 500**

Attempting to create a meeting without any invitee or removing all invitees from the meeting will lead to an error

- [x] **Discrepancy between invitees and responsible selection**

Users not selected as invitees can still be selected as being responsible for an agenda item.

<img width="585" alt="image" src="https://github.com/opf/openproject/assets/617519/39217c44-ace8-4404-88f5-f55a8cacfae4">

Even members without permissions to see the meeting can be selected. Those users are correctly removed from the list possible participants.

- [x] **Typing in the responsible selection field leads to form resizing**

When typing in the responsible selection autocompleter the field changes its size due to the loading animation leading to the whole form being resized.

- [x] **Focus misplaced when adding work package items**

After pressing the "Add" button and selecting to want to add a "Work package", the focus is in the responsible autocompleter instead of in the title field.

<img width="1009" alt="image" src="https://github.com/opf/openproject/assets/617519/67752403-d2f8-4edd-b531-8050590efd4e">

- [x] **Adding a large value for the duration crashes the application**

A large value for the duration field crashes the application:

<img width="1018" alt="image" src="https://github.com/opf/openproject/assets/617519/16c7d980-508a-484a-98b5-84ed28304c4e">

Adding a validation for the upper boundary (and possibly also limiting the field visually) would help.

- [x] **Opinion/Feedback: Always show the description field**

Having to click on the '+ Description' button to display an input is cumbersome. Clicking on the button will only display a single field (and I cannot hide it again). Given what it currently is, I'd always display the field. If the assumption is that people will only rarely insert a description  this might not be necessary but I believe, that they will insert it more often especially if they want to take notes of the result. And in that case, they have to click two buttons while being in a meeting to take minutes per item.

Decision with Parimal: Description will stay as it is

- [x] **Opinion/Feedback: Naming of the date field**

The field when editing the meeting's attributes is named "Start date" suggesting that there would also be a finish date

<img width="465" alt="image" src="https://github.com/opf/openproject/assets/617519/4f3a36d8-f921-4481-be2a-b4a5aaf02158">

Admittedly, it is called the same in the create form but there the date is combined with the time which to me makes it a little better.

Decision: We'll rename the field to "Time"

- [x] **Two ways of configuring the participants in close proximity**

Clicking on the gear icon as well as clicking on 'Add participants' triggers the same modal and the two options are in close proximity to each other.

Decision: Will stay as it is after discussion with Parimal

- [x] **Opinon/Feedback: Change default of meeting type**

Instead of having the old style of meeting selected by default I'd chose the new one:

<img width="1031" alt="image" src="https://github.com/opf/openproject/assets/617519/d5edf449-a8fa-42e1-865f-2723c298f8c4">

- [x] **Drag & Drop while editing/creating an agenda item looses user input**

When draging&dropping and dropping while having an agenda item open to be created/edited will lead to completely rerendering the agenda section potentially loosing text the user has created. It might be an option to disable drag&drop while editing.   

To be clarified

- [x] **Deleting a structured meeting leads to a white page**

<img width="800" alt="image" src="https://github.com/opf/openproject/assets/617519/b3fbdf95-f1d4-4527-b8ec-5ae3d1e3b48a">

- [x] **Opinion/Feedback: Added content is not scrolled into view**

When having a longer meeting agenda and adding another one, that new item is not scrolled completely into view making editing awkward:

<img width="1270" alt="image" src="https://github.com/opf/openproject/assets/617519/8cb2d318-d452-4086-af32-991ab5fcf52d">

- [x] Height and width of autocompleter components change / are not consistent

![Bildschirmfoto 2023-09-18 um 08 53 40](https://github.com/opf/openproject/assets/459462/e8349e86-6640-4ecf-b1a6-54eba6aa7e7d)

oliverguenther committed:
Fix test being too fast inserting two titles

ulferts commented on app/forms/op_primer/forms/autocompleter.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module OpPrimer
```
Why is this situated in `app/forms` when the rest of the custom components are in `app/components`? It does now have `forms` in its path twice.

cbliard replied:
[Current recommendation](https://primer.style/view-components/lookbook/pages/forms/getting_started) is to put form objects in `app/forms` and mimic the same directory structure as `app/controllers`.

These files are not forms, but rather extensions of the forms dsl. They have been moved to `lib/` in 0a4dbe97a. But once there are forms, they'll probably live in `app/forms` anyway. For the meeting module, they live in `modules/meeting/app/forms/meeting`.

ulferts replied:
Understood, @cbliard. Now, those auxiliary form files are in lib/primer/open_project and the actual forms used in pages will be in `app/forms`. I can understand that distinction. Thanks for clarifying.  

I like that the auxiliary classes now are also in `Primer::OpenProject` same as the components by which we extend `Primer`.

oliverguenther committed:
Don't try to access author if none set

ulferts commented on app/views/homescreen/index.html.erb:
```diff
@@ -58,4 +61,6 @@ See COPYRIGHT and LICENSE files for more details.
 <% end %>
 
+
+
```
Those changes are unnecessary

cbliard replied:
There are editor configs to do it automatically:

* VSCode: `files.trimFinalNewlines: true`
* Atom: `whitespace.ensureSingleTrailingNewline: true`
* RubyMine: under Settings > Editor > General
  * check "Remove traling blank lines at the end of saved files"
  * check "Ensure every saved file ends with a line break"

oliverguenther committed:
Validate duration

ulferts commented on frontend/src/global_styles/openproject/_primer-adjustments.sass:
```diff
@@ -0,0 +1,44 @@
+//++
+
+// Required as the OpenProject form styles override the Primer form styles
+// disabled for now in order to see the conflicts
```
So this file is responsible for reasserting the styles already defined in Primer by a higher specificity? It does not override the primer styles? If so, the comment could be a bit clearer on that.

cbliard replied:
There are editor configs to do it automatically:

* VSCode: `files.trimFinalNewlines: true`
* Atom: `whitespace.ensureSingleTrailingNewline: true`
* RubyMine: under Settings > Editor > General
  * check "Remove traling blank lines at the end of saved files"
  * check "Ensure every saved file ends with a line break"

oliverguenther committed:
Fix deleting meetings without turbo

ulferts commented on frontend/src/stimulus/controllers/dynamic/meeting-agenda-item-form.controller.ts:
```diff
@@ -0,0 +1,90 @@
+      const text = await response.text();
+      Turbo.renderStreamMessage(text);
+    }
+  }
```
I think this isn't actually used (any more?).

ulferts commented on modules/meeting/app/components/flash_message_component.rb:
```diff
@@ -0,0 +1,62 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class FlashMessageComponent < ApplicationComponent
```
Since the intend of this is to be used throughout the application, it should be moved to OpPrimer::FlashMessageComponent or Primer::OpenProject::FlashMessageComponent.

cbliard committed:
Rename OpPrimer::Forms to Primer::OpenProject::Forms

This is done to match the namespace used in `opf/primer_view_components`
which is `Primer::OpenProject` for our other custom components.

ulferts commented on modules/meeting/app/components/meeting_agenda_items/item_component/show_component.rb:
```diff
@@ -0,0 +1,272 @@
+    end
+
+    def edit_enabled?
+      @meeting.open? && User.current.allowed_to?(:create_meeting_agendas, @meeting.project)
```
The mapping of the already existing permissions to the new paradigm doesn't work very well. The agenda items are edited also during the meeting as part of the minutes. 

We could introduce new permissions that better reflect the altered paradigms and have the old permissions faded out. Or we simply use the `:edit_meetings` permission for all of those checks.

cbliard commented on modules/meeting/app/forms/meeting_agenda_item/duration.rb:
```diff
@@ -34,6 +34,7 @@ class MeetingAgendaItem::Duration < ApplicationForm
       label: MeetingAgendaItem.human_attribute_name(:duration_in_minutes),
       leading_visual: { icon: :stopwatch },
       visually_hide_label: true,
+      max: 86400,
```
That's 60 days. Is it intended? Did you mean `24*60` instead? (`24.hours.in_minutes.to_i`)

oliverguenther committed:
Fix submission of forms in CKEditor with turbo

The CKEditor form wrapper cancels and triggers resubmission of the form to be able to get all data from CKEditor
(which is asynchronous so the latest input from the user might not be present).

However turbo wraps submission events through clicks on submit and other events, so if we cancel the first submit,
the second submit will not go through turbo.

To ensure we can resubmit, we need to use turbo's own navigator, which again breaks on non-turbo pages.

oliverguenther committed:
fixup! Validate duration

oliverguenther committed:
fixup! Fix submission of forms in CKEditor with turbo

oliverguenther commented:
I think I've caught all the bugs you mentioned, we'll discuss the feedback next week

> Possible to create an agenda item without user

This is by design and should not break (fixed)

> Removing the last invitee from the meeting leads to 500

Couldn't reproduce this @ulferts , does it happen on the head of the branch to you?

> Adding a large value for the duration crashes the application

Fixed (arbitrary max value of 1440)

> Focus misplaced when adding work package items

Should be fixed already in the meantime

> Drag & Drop while editing/creating an agenda item looses user input

this was mentioned by Jonas already as a know caveat, we'll discuss this with product

> Deleting a structured meeting leads to a white page

Fixed, caused by turbo taking over as not explicitly disabled

ulferts commented on modules/meeting/app/components/meeting_agenda_items/list_component.rb:
```diff
@@ -0,0 +1,113 @@
+            end
+          end
+        end
+        # component.with_primary_action(href: "#").with_content(t("label_meeting_empty_action"))
```
Leftover?

oliverguenther committed:
Replace avatar rendered in primer with opce-principal

ulferts commented on modules/meeting/app/components/meetings/sidebar/details_component.rb:
```diff
@@ -0,0 +1,176 @@
+      duration = Duration.new(seconds: @meeting.duration * 3600)
+
+      meeting_attribute_row(:stopwatch) do
+        if duration.hours > 0
+          render(Primer::Beta::Text.new) do
+            "#{duration.hours} h #{duration.minutes} min"
+          end
+        else
+          render(Primer::Beta::Text.new) do
+            "#{duration.minutes} min"
+          end
+        end
+      end
```
This part should be extracted into a component for rendering durations. We are doing it on the meetings page two times already (the other usage is in the `MeetingAgendaItems::ItemComponent::ShowComponent`) and I'd expect we do it again.

Interestingly enough I couldn't find good information on how durations are internationalized but having the format moved to our yml files sound good to me.

ulferts reviewed (changes requested):
I did not manage to go through the whole PR yet and plan to continue tomorrow. Please find most of my remarks next to the code.

There are also a couple of general issues

- [x] **Members autocompleter broken**

(OG: Cannot reproduce, are you sure?) (JU: I had some invalid data in the database I was testing with. Sorry for the confusion)


The members autocompleter has become non functional. The fact that the specs did not catch it is unnerving.

<img width="741" alt="image" src="https://github.com/opf/openproject/assets/617519/78e6fe2f-cbba-407e-ae88-658d3fbac0d5">


- [x] Avatars on members page no longer have a padding

![Bildschirmfoto 2023-09-18 um 09 09 29](https://github.com/opf/openproject/assets/459462/ac54faa2-e72a-438e-9682-d7cec517f763)



- [x] **Non working days in primer date selector**

(OG: This is a browser native datepicker)

The date selector of primer naturally does not display the non working days:

<img width="236" alt="image" src="https://github.com/opf/openproject/assets/617519/98e46f6b-31fe-4ce1-b9ea-7c6a6165d5e5">

Is there any chance we can get them to be adhered to?

ulferts commented:
>  
> > Removing the last invitee from the meeting leads to 500
> 
> Couldn't reproduce this @ulferts , does it happen on the head of the branch to you?
> 

It's now working without problem @oliverguenther

ulferts commented:
Another feedback @oliverguenther. I believe that the participant modal can become unwieldy if there are a larger number of members (e.g. 100) in the group. Admittedly, a problem inherited from the existing participant selection.

ulferts commented on modules/meeting/app/components/meeting_agenda_items/item_component/show_component.rb:
```diff
@@ -0,0 +1,268 @@
+               style: "cursor: move;",
+               size: :small,
+               icon: :grabber
+             ))
```
I am rather sure that a `DragHandlerComponent` will be used again and again, so I'd rather extract it now.

ulferts commented on modules/meeting/app/components/meeting_agenda_items/item_component/show_component.rb:
```diff
@@ -0,0 +1,268 @@
+            @meeting_agenda_item.work_package.subject
+          end
+        end
+      end
```
This part could also be extracted into a component.

ulferts commented on modules/meeting/app/components/meeting_agenda_items/item_component/show_component.rb:
```diff
@@ -0,0 +1,268 @@
+
+    def show_time_slot?
+      false
+    end
```
Unless this is planned to be used, please remove the method and those affected by it, e.g. `time_slot_partial`

ulferts commented on modules/meeting/app/components/meeting_agenda_items/new_component.rb:
```diff
@@ -0,0 +1,69 @@
+      super
+
+      @meeting = meeting
+      @meeting_agenda_item = meeting_agenda_item || MeetingAgendaItem.new(meeting:, author: User.current)
```
The initialization of a new item should rather be done in the controller (where we formerly relied on the SetAttributesService to do so) than in the view.

oliverguenther replied:
The problem here is that this is called from multiple places, so you'd need to duplicate the initialization logic. I suggest we discuss a convention for this separately

ulferts commented on modules/meeting/app/components/meetings/header_component.rb:
```diff
@@ -0,0 +1,195 @@
+#++
+
+module Meetings
+  class HeaderComponent < ApplicationComponent
```
We do have the `Primer::OpenProject::PageHeader` component in by now and this looks to be a good case where it should be applied - at least for the show state.

oliverguenther replied:
We will replace the header component with the PageHeader as soon as it becomes editable

ulferts commented on modules/meeting/app/components/meetings/row_component.rb:
```diff
@@ -38,6 +38,14 @@ def title
+      if model.is_a?(StructuredMeeting)
+        I18n.t('meeting.types.structured')
+      else
+        I18n.t('meeting.types.classic')
```
A key other then classic would be more descriptive

ulferts commented on modules/meeting/app/components/meetings/sidebar_component.rb:
```diff
@@ -0,0 +1,71 @@
+    def participants_partial
+      render(Meetings::Sidebar::ParticipantsComponent.new(meeting: @meeting))
+    end
+  end
```
It appears that we are rebuilding GitHub's sidebar here which is not part of Primer. 

When looking at the CSS classes we seem to be doing it a way different from the way GitHub does it even though the result looks the same. 

We should move the component to a more general place so that it can be reused, ideally into the `primer_view_components` repo. To me, it would also make sense to mimic the styles used by GitHub in the implementation.

ulferts commented on modules/meeting/app/components/users/avatar_component.rb:
```diff
@@ -0,0 +1,54 @@
+
+    def call
+      helpers.avatar(@user, size: @size, hide_name: !@show_name)
+    end
```
This component isn't meeting specific and should be moved to `/app/components`

ulferts commented on modules/meeting/app/contracts/meeting_agenda_items/delete_contract.rb:
```diff
@@ -0,0 +1,41 @@
+        errors.add :base, I18n.t(:text_meeting_not_editable_anymore)
+      end
+    end
+  end
```
This contract does not check permissions. Is this intended as the permissions are checked elsewhere?

If permissions are to be checked. Inheriting from `DeleteContract` would make that easier.

ulferts commented on modules/meeting/app/contracts/meeting_agenda_items/create_contract.rb:
```diff
@@ -0,0 +1,41 @@
+    def validate_editable
+      unless model.editable?
+        errors.add :base, I18n.t(:text_meeting_not_editable_anymore)
+      end
```
No permission check is applied here. Is this intended and is the permission checked elsewhere?

ulferts commented on modules/meeting/app/contracts/meeting_agenda_items/update_contract.rb:
```diff
@@ -0,0 +1,41 @@
+      unless model.editable?
+        errors.add :base, I18n.t(:text_meeting_not_editable_anymore)
+      end
+    end
```
Again, no permission checked. And the contents is basically the same as in the other contracts so using a module could reduce duplication.

ulferts commented on modules/meeting/app/controllers/concerns/agenda_component_streams.rb:
```diff
@@ -0,0 +1,148 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module AgendaComponentStreams
```
Namespacing this so that in reads `Meetings::AgendaComponentStreams` would make the references in the code clearer.

ulferts commented on app/controllers/concerns/op_turbo/component_stream.rb:
```diff
@@ -0,0 +1,79 @@
+
+    def prepend_via_turbo_stream(component:, target_component:)
+      @turbo_streams << target_component.insert_as_turbo_stream(component:, view_context:, action: :prepend)
+    end
```
the before action and explicit initialization of the `@turbo_streams` variable is not really needed if a method like this is introduced

```
def turbo_streams
  @turbo_streams ||= []
end
```

and only the method is called.

ulferts reviewed (changes requested):
I still did not manage to work through all of the PR and plan to continue to do so on Monday. Most of the remarks up until now are at the code but some is also more general:

- [x] **CSS included multiple times:**

The css seems to be included multiple times:

<img width="514" alt="image" src="https://github.com/opf/openproject/assets/617519/eddb7c36-f516-471e-92de-9feb5fdfd88d">

I did not check whether that is a problem originating in this PR.

- [x] **Empty page on canceling the editing of the meeting title**

Clicking on 'Cancel' when editing the meeting title leads to an empty page:

<img width="620" alt="image" src="https://github.com/opf/openproject/assets/617519/5e83b5e2-44b1-43a9-9ba4-0b4a64a005fb">
 
- [ ] **Opinion/Feedback: Closing (the agenda) and editing other values**

It feels weird that I can close/lock the agenda and then still edit the meetings attributes and participants:

Discussion with Parimal: Disable editing of participants, details, title when in closed mode

<img width="191" alt="image" src="https://github.com/opf/openproject/assets/617519/0dca4a23-9e97-4b65-9b99-88ef00a83a19">

Especially while we have no built in support for the minutes which will probably result in people writing that information inside the "Description" field, closing (the agenda) will, if at all , be done at the end of the meeting. And then, the other attribute should be locked as well.

oliverguenther committed:
replace before_action with memoized method

oliverguenther committed:
Namespace AgendaComponentStreams

oliverguenther committed:
Remove show_time_slot

oliverguenther committed:
Remove duplicate import of primer/css

oliverguenther commented:
> CSS included multiple times:

Not introduced in this PR: https://github.com/opf/openproject/pull/13720

oliverguenther committed:
Remove trailing lines and spaces

oliverguenther committed:
Remove cancel method/value

oliverguenther commented:
- [ ] When creating a new agenda item, selecting a user and leaving the title blank, then saving the agenda item "resets" the user selection hidden input. When you fill out a title and hit save, a different user is saved.

oliverguenther committed:
Scroll new agenda item into view

oliverguenther committed:
Clarify comment in primer-adjustments

oliverguenther committed:
Move avatar component and make preview

oliverguenther commented:
- [x] Escape on title input should cancel create and editing of an meeting agenda item (no confirmation)
- [x] Change description to notes

oliverguenther committed:
Rename Description => Notes

oliverguenther committed:
Restore escape on title -> cancel editing

oliverguenther committed:
Rewrite spec without cuprite due to keyboard bug

see https://github.com/rubycdp/cuprite/issues/240

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
Remove author field from meeting agenda

oliverguenther committed:
Add title to author

oliverguenther committed:
Make author mandatory

oliverguenther committed:
Fix primerized height

oliverguenther committed:
Change avatar + manual name to component

HDinger committed:
Move font-size setting to body. In 0.875 rem inside the html would reduce the base reference font size for all other rem calculations resultung in a general font size of 12.5px

oliverguenther committed:
Fix avatar expectation

oliverguenther committed:
Fix expected string

oliverguenther committed:
Validate presence of author

oliverguenther committed:
Move permission checks to edit_meetings

jjabari-op committed:
added missing permission mapping for the cancel edit action

jjabari-op committed:
Merge branch 'epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages' of github.com:opf/openproject into epic/37297-automatic-meeting-agendas-and-minutes-linked-to-work-packages

oliverguenther committed:
Extract editable check into BaseContract

oliverguenther committed:
Add specs for contracts

oliverguenther commented:
Squashed into https://github.com/opf/openproject/pull/13725
